src/wallet/wallet.cpp

```diff
@@ -2733,6 +2733,28 @@ bool CWallet::AttachChain(const std::shared_ptr<CWallet>& walletInstance, interf
     assert(!walletInstance->m_chain || walletInstance->m_chain == &chain);
     walletInstance->m_chain = &chain;
 
+    // Unless allowed, ensure wallet files are not reused across chains:
+    if (!gArgs.GetBoolArg("-walletcrosschain", DEFAULT_WALLETCROSSCHAIN))
+    {
+        WalletBatch batch(walletInstance->GetDatabase());
+        CBlockLocator locator;
+        if (batch.ReadBestBlock(locator))
+        {
+            // Wallet is assumed to be from another chain, if none of
+            // its last 6 best known blocks are in the active chain
+            // (this heavily relies on the fact that locator stores
+            // last 10 blocks consecutively):
+            const std::optional<int> fork_height = chain.findLocatorFork(locator);
+            if (fork_height && std::distance(locator.vHave.begin(),
+                                             std::find(locator.vHave.begin(), locator.vHave.end(),
+                                                       chain.getBlockHash(*fork_height))) > 6)
+            {
+                error = _("Wallet files should not be reused across chains.");
+                return false;
+            }
+        }
+    }
+
```

Comment:
Had to move it here, since putting this after line 2766 where I placed it originally results in memory leaks, as detected by the sanitizer.